### PR TITLE
Adds a system check for fips enablement and runs in fips mode if enabled

### PR DESCRIPTION
**Pull request by mwrock**
This essentially refactors code from #4841 and https://github.com/chef/ohai/pull/803 into `ChefConfig` the next step will be to refactor ohai code out of ohai and consume this instead.

### CODE REVIEW

File: chef-config/lib/chef-config/fips.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        end
+      rescue Win32::Registry::Error
+        false
+      end
+    else
+      begin
+        File.read("/proc/sys/crypto/fips_enabled").chomp == "0" ? false : true
+      rescue Errno::ENOENT
+        false
+      end
```
Comment (lamont-granquist) [inline chef-config/lib/chef-config/fips.rb]: could probably be:

```
File.exist?("/proc/sys/crypto/fips_enabled") && File.read("/proc/sys/crypto/fips_enabled").chomp != "0"
```

avoid using the exception as normal flow control and invert the conditional...

File: chef-config/lib/chef-config/fips.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    if ChefConfig.windows?
+      require "win32/registry"
+      # from http://msdn.microsoft.com/en-us/library/windows/desktop/aa384129(v=vs.85).aspx
+      if ::RbConfig::CONFIG["target_cpu"] == "i386"
+        reg_type = Win32::Registry::KEY_READ | 0x100
+      elsif ::RbConfig::CONFIG["target_cpu"] == "x86_64"
+        reg_type = Win32::Registry::KEY_READ | 0x200
+      else
+        reg_type = Win32::Registry::KEY_READ
+      end
```
Comment (lamont-granquist) [inline chef-config/lib/chef-config/fips.rb]:
```
reg_type =
  case ::RbConfig::CONFIG["target_cpu"]
  when "i386"
    Win32::Registry::KEY_READ | 0x100
  when "x86_64"
    Win32::Registry::KEY_READ | 0x200
  else 
    Win32::Registry::KEY_READ
  end
```

File: chef-config/lib/chef-config/fips.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+      begin
+        Win32::Registry::HKEY_LOCAL_MACHINE.open('System\CurrentControlSet\Control\Lsa\FIPSAlgorithmPolicy', reg_type) do |policy|
+          policy["Enabled"] == 0 ? false : true
```
Comment (lamont-granquist) [inline chef-config/lib/chef-config/fips.rb]: `policy["Enabled"] != 0`

**Comment by lamont-granquist** [comment]
👍  so much better

File: chef-config/lib/chef-config/fips.rb (hunk 3 above)
Comment (mwrock) [inline chef-config/lib/chef-config/fips.rb]: ugh yeah. I dont know what i was thinking with the `?` `:` here and above

Comment (lamont-granquist) [inline chef-config/lib/chef-config/fips.rb]: been a long week and its time for beer-thirty?

**Comment by thommay** [comment]
👍

**Comment by mwrock** [comment]
just working on getting externable tests in an agreable state and I'll merge.

**Pull request merged**